# IMDB Big Data Analysis - Complete Project

This notebook contains the complete analysis of IMDB datasets, answering all 14 questions.

**Team Members:**
- Samuel Giorno
- Maxime Boiral
- Lou Bruneau


## Setup and Imports

In [ ]:
# Import required libraries
import sys
import pandas as pd
import numpy as np
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

# Add src directory to path for importing custom modules
sys.path.append('../src')
from data_loader import IMDBDataLoader

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)
pd.set_option('display.width', 1000)

print("Libraries imported successfully!")
print(f"Pandas version: {pd.__version__}")
print(f"NumPy version: {np.__version__}")

Libraries imported successfully!
Pandas version: 2.3.3
NumPy version: 1.26.4


## Question 1: Load Data from IMDB

We will download and load the necessary datasets from https://datasets.imdbws.com/

The following datasets will be used:
- **name.basics.tsv.gz** - Information about people (actors, directors, etc.)
- **title.basics.tsv.gz** - Basic information about titles (movies, TV shows, etc.)
- **title.ratings.tsv.gz** - Ratings for titles
- **title.crew.tsv.gz** - Directors and writers for titles
- **title.akas.tsv.gz** - Alternative titles/names

In [ ]:
# Initialize the data loader
loader = IMDBDataLoader(data_dir='../data') 

# Check what datasets we already have
loader.print_dataset_info()


IMDB Dataset Information
Dataset              Filename                  Status     Size (MB)
----------------------------------------------------------------------
name_basics          name.basics.tsv.gz        ✓ Found    281.3
title_basics         title.basics.tsv.gz       ✓ Found    204.2
title_ratings        title.ratings.tsv.gz      ✓ Found    7.7
title_crew           title.crew.tsv.gz         ✓ Found    75.2
title_akas           title.akas.tsv.gz         ✓ Found    448.4
title_principals     title.principals.tsv.gz   ✗ Missing  -
title_episode        title.episode.tsv.gz      ✗ Missing  -



In [3]:
print("Starting downloads...\n")

datasets_to_download = [
    'name_basics',
    'title_basics',
    'title_ratings',
    'title_crew',
    'title_akas'
]

for dataset in datasets_to_download:
    print(f"\n{'='*60}")
    print(f"Downloading: {dataset}")
    print('='*60)
    loader.download_dataset(dataset)

print("\n" + "="*60)
print("All downloads complete!")
print("="*60)

Starting downloads...


Downloading: name_basics
✓ name.basics.tsv.gz already exists, skipping download

Downloading: title_basics
✓ title.basics.tsv.gz already exists, skipping download

Downloading: title_ratings
✓ title.ratings.tsv.gz already exists, skipping download

Downloading: title_crew
✓ title.crew.tsv.gz already exists, skipping download

Downloading: title_akas
✓ title.akas.tsv.gz already exists, skipping download

All downloads complete!


### Load Datasets into Memory

Now we'll load the datasets into pandas DataFrames. For very large datasets, we may need to use chunked loading or column selection to manage memory.

In [4]:
# Load name.basics (people data)
print("Loading name.basics...")
df_names = loader.load_dataset('name_basics')

# Display first few rows and basic info
print("\nFirst 5 rows:")
display(df_names.head())
print("\nDataset info:")
print(df_names.info())

Loading name.basics...
Loading name.basics.tsv.gz...
✓ Loaded 14,901,541 rows and 6 columns in 31.2s
  Memory usage: 4100.7MB

First 5 rows:


,nconst,primaryName,birthYear,deathYear,primaryProfession,knownForTitles
0,nm0000001,Fred Astaire,1899.0,1987.0,"actor,miscellaneous,producer","tt0072308,tt0050419,tt0027125,tt0025164"
1,nm0000002,Lauren Bacall,1924.0,2014.0,"actress,miscellaneous,soundtrack","tt0037382,tt0075213,tt0038355,tt0117057"
2,nm0000003,Brigitte Bardot,1934.0,NaN,"actress,music_department,producer","tt0057345,tt0049189,tt0056404,tt0054452"
3,nm0000004,John Belushi,1949.0,1982.0,"actor,writer,music_department","tt0072562,tt0077975,tt0080455,tt0078723"
4,nm0000005,Ingmar Bergman,1918.0,2007.0,"writer,director,actor","tt0050986,tt0069467,tt0050976,tt0083922"



Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14901541 entries, 0 to 14901540
Data columns (total 6 columns):
 #   Column             Dtype  
---  ------             -----  
 0   nconst             object 
 1   primaryName        object 
 2   birthYear          float64
 3   deathYear          float64
 4   primaryProfession  object 
 5   knownForTitles     object 
dtypes: float64(2), object(4)
memory usage: 682.1+ MB
None


In [5]:
# Load title.basics (title information)
print("Loading title.basics...")
df_titles = loader.load_dataset('title_basics')

print("\nFirst 5 rows:")
display(df_titles.head())
print("\nDataset info:")
print(df_titles.info())

Loading title.basics...
Loading title.basics.tsv.gz...
✓ Loaded 12,093,921 rows and 9 columns in 27.5s
  Memory usage: 4847.0MB

First 5 rows:


,tconst,titleType,primaryTitle,originalTitle,isAdult,startYear,endYear,runtimeMinutes,genres
0,tt0000001,short,Carmencita,Carmencita,0,1894.0,NaN,1,"Documentary,Short"
1,tt0000002,short,Le clown et ses chiens,Le clown et ses chiens,0,1892.0,NaN,5,"Animation,Short"
2,tt0000003,short,Poor Pierrot,Pauvre Pierrot,0,1892.0,NaN,5,"Animation,Comedy,Romance"
3,tt0000004,short,Un bon bock,Un bon bock,0,1892.0,NaN,12,"Animation,Short"
4,tt0000005,short,Blacksmith Scene,Blacksmith Scene,0,1893.0,NaN,1,Short



Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12093921 entries, 0 to 12093920
Data columns (total 9 columns):
 #   Column          Dtype  
---  ------          -----  
 0   tconst          object 
 1   titleType       object 
 2   primaryTitle    object 
 3   originalTitle   object 
 4   isAdult         int64  
 5   startYear       float64
 6   endYear         float64
 7   runtimeMinutes  object 
 8   genres          object 
dtypes: float64(2), int64(1), object(6)
memory usage: 830.4+ MB
None


In [6]:
# Load title.ratings (ratings data)
print("Loading title.ratings...")
df_ratings = loader.load_dataset('title_ratings')

print("\nFirst 5 rows:")
display(df_ratings.head())
print("\nDataset info:")
print(df_ratings.info())

Loading title.ratings...
Loading title.ratings.tsv.gz...
✓ Loaded 1,605,268 rows and 3 columns in 1.8s
  Memory usage: 126.0MB

First 5 rows:


,tconst,averageRating,numVotes
0,tt0000001,5.7,2186
1,tt0000002,5.5,306
2,tt0000003,6.4,2270
3,tt0000004,5.2,196
4,tt0000005,6.2,3010



Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1605268 entries, 0 to 1605267
Data columns (total 3 columns):
 #   Column         Non-Null Count    Dtype  
---  ------         --------------    -----  
 0   tconst         1605268 non-null  object 
 1   averageRating  1605268 non-null  float64
 2   numVotes       1605268 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 36.7+ MB
None


In [7]:
# Load title.crew (directors and writers)
print("Loading title.crew...")
df_crew = loader.load_dataset('title_crew')

print("\nFirst 5 rows:")
display(df_crew.head())
print("\nDataset info:")
print(df_crew.info())

Loading title.crew...
Loading title.crew.tsv.gz...
✓ Loaded 12,095,664 rows and 3 columns in 14.8s
  Memory usage: 2027.6MB

First 5 rows:


,tconst,directors,writers
0,tt0000001,nm0005690,NaN
1,tt0000002,nm0721526,NaN
2,tt0000003,nm0721526,nm0721526
3,tt0000004,nm0721526,NaN
4,tt0000005,nm0005690,NaN



Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12095664 entries, 0 to 12095663
Data columns (total 3 columns):
 #   Column     Dtype 
---  ------     ----- 
 0   tconst     object
 1   directors  object
 2   writers    object
dtypes: object(3)
memory usage: 276.8+ MB
None


In [8]:
# Load title.akas (alternative titles)
print("Loading title.akas...")
df_akas = loader.load_dataset('title_akas')

print("\nFirst 5 rows:")
display(df_akas.head())
print("\nDataset info:")
print(df_akas.info())

Loading title.akas...
Loading title.akas.tsv.gz...
✓ Loaded 53,992,154 rows and 8 columns in 64.4s
  Memory usage: 17900.3MB

First 5 rows:


,titleId,ordering,title,region,language,types,attributes,isOriginalTitle
0,tt0000001,1,Carmencita,NaN,NaN,original,NaN,1
1,tt0000001,2,Carmencita,DE,NaN,NaN,literal title,0
2,tt0000001,3,Carmencita,US,NaN,imdbDisplay,NaN,0
3,tt0000001,4,Carmencita - spanyol tánc,HU,NaN,imdbDisplay,NaN,0
4,tt0000001,5,Καρμενσίτα,GR,NaN,imdbDisplay,NaN,0



Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53992154 entries, 0 to 53992153
Data columns (total 8 columns):
 #   Column           Dtype 
---  ------           ----- 
 0   titleId          object
 1   ordering         int64 
 2   title            object
 3   region           object
 4   language         object
 5   types            object
 6   attributes       object
 7   isOriginalTitle  int64 
dtypes: int64(2), object(6)
memory usage: 3.2+ GB
None


### Question 1 - Answer

✅ **Data successfully loaded programmatically from IMDB datasets.**

All required datasets have been downloaded and loaded into memory:
- name.basics.tsv.gz
- title.basics.tsv.gz
- title.ratings.tsv.gz
- title.crew.tsv.gz
- title.akas.tsv.gz

## Question 2: How many total people in dataset?

In [9]:
total_people = len(df_names)

print(f"Total number of people in dataset: {total_people:,}")
print(f"\nUnique identifiers (nconst): {df_names['nconst'].nunique():,}")
print(f"Duplicate entries: {total_people - df_names['nconst'].nunique()}")

Total number of people in dataset: 14,901,541

Unique identifiers (nconst): 14,901,541
Duplicate entries: 0


### Q2 - Answer

**Total people in dataset:** 14,901,541

## Question 3: What is the earliest year of birth?

In [10]:
# Convert birthYear to numeric (it may be stored as string)
df_names['birthYear'] = pd.to_numeric(df_names['birthYear'], errors='coerce')

# Find minimum birth year (excluding NaN)
earliest_birth_year = df_names['birthYear'].min()

# Get the person(s) with this birth year
earliest_person = df_names[df_names['birthYear'] == earliest_birth_year]

print(f"Earliest year of birth: {int(earliest_birth_year)}")
print(f"\nPerson(s) with this birth year:")
display(earliest_person[['nconst', 'primaryName', 'birthYear', 'deathYear', 'primaryProfession', 'knownForTitles']])

Earliest year of birth: 4

Person(s) with this birth year:


,nconst,primaryName,birthYear,deathYear,primaryProfession,knownForTitles
737971,nm0784172,Lucio Anneo Seneca,4.0,65.0,writer,"tt0043802,tt0218822,tt0049203,tt0972562"


### Question 3 - Answer

**Earliest year of birth:** 4

This represents the oldest birth year recorded in the IMDB dataset.

## Question 4: How many years ago was this person born?

In [11]:
# Calculate how many years ago the person was born
current_year = datetime.now().year
years_ago = current_year - int(earliest_birth_year)

print(f"Current year: {current_year}")
print(f"Birth year: {int(earliest_birth_year)}")
print(f"\nYears ago: {years_ago} years")

Current year: 2025
Birth year: 4

Years ago: 2021 years


### Question 4 - Answer

**Years ago:** 2021

## Question 5 & 6: Validate earliest birth date and explain reasoning

In [12]:
# Get detailed information about the person with earliest birth year
person_id = earliest_person.iloc[0]['nconst']
person_name = earliest_person.iloc[0]['primaryName']
birth_year = int(earliest_person.iloc[0]['birthYear'])
death_year = earliest_person.iloc[0]['deathYear']
known_titles = earliest_person.iloc[0]['knownForTitles']

print(f"Analyzing: {person_name} (ID: {person_id})")
print(f"Birth Year: {birth_year}")
print(f"Death Year: {death_year if pd.notna(death_year) else 'Not recorded'}")
print(f"\nKnown for titles: {known_titles}")

# If person has death year, calculate lifespan
if pd.notna(death_year):
    death_year_int = int(death_year)
    lifespan = death_year_int - birth_year
    print(f"\nCalculated lifespan: {lifespan} years")
    print(f"Death year: {death_year_int}")

# Look up the titles this person is known for
if pd.notna(known_titles) and known_titles != '':
    title_ids = known_titles.split(',')
    print(f"\nAnalyzing {len(title_ids)} known titles...")
    
    # Get information about these titles
    known_works = df_titles[df_titles['tconst'].isin(title_ids)]
    
    if len(known_works) > 0:
        print("\nKnown works:")
        display(known_works[['tconst', 'primaryTitle', 'titleType', 'startYear', 'endYear', 'runtimeMinutes', 'genres']])
        
        # Calculate ages at time of work
        known_works_copy = known_works.copy()
        known_works_copy['startYear'] = pd.to_numeric(known_works_copy['startYear'], errors='coerce')
        known_works_copy['age_at_work'] = known_works_copy['startYear'] - birth_year
        
        print("\nAge at time of each work:")
        for _, work in known_works_copy.iterrows():
            if pd.notna(work['age_at_work']):
                print(f"  {work['primaryTitle']} ({int(work['startYear'])}): Age {int(work['age_at_work'])}")
        
        min_age = known_works_copy['age_at_work'].min()
        max_age = known_works_copy['age_at_work'].max()
        
        if pd.notna(min_age) and pd.notna(max_age):
            print(f"\nWorking age range: {int(min_age)} to {int(max_age)} years old")

Analyzing: Lucio Anneo Seneca (ID: nm0784172)
Birth Year: 4
Death Year: 65.0

Known for titles: tt0043802,tt0218822,tt0049203,tt0972562

Calculated lifespan: 61 years
Death year: 65

Analyzing 4 known titles...

Known works:


,tconst,primaryTitle,titleType,startYear,endYear,runtimeMinutes,genres
43026,tt0043802,The Affairs of Messalina,movie,1951.0,NaN,116,"Action,Drama,History"
48311,tt0049203,"Fedra, the Devil's Daughter",movie,1956.0,NaN,100,Drama
209696,tt0218822,Such Is Life,movie,2000.0,NaN,98,"Drama,Romance"
940991,tt0972562,Medea 2,movie,2006.0,NaN,NaN,Drama



Age at time of each work:
  The Affairs of Messalina (1951): Age 1947
  Fedra, the Devil's Daughter (1956): Age 1952
  Such Is Life (2000): Age 1996
  Medea 2 (2006): Age 2002

Working age range: 1947 to 2002 years old


In [13]:
# Additional validation: Check for statistical anomalies
print("=" * 70)
print("VALIDATION ANALYSIS")
print("=" * 70)

# 1. Check distribution of birth years
birth_year_stats = df_names['birthYear'].describe()
print("\nBirth year statistics:")
print(birth_year_stats)

# 2. Find all people born in the same century
century = (birth_year // 100) * 100
same_century = df_names[(df_names['birthYear'] >= century) & (df_names['birthYear'] < century + 100)]
print(f"\nPeople born in {century}s: {len(same_century):,}")

# 3. Check oldest 10 people
oldest_10 = df_names.nsmallest(10, 'birthYear')[['primaryName', 'birthYear', 'deathYear', 'primaryProfession']]
print("\nOldest 10 people in dataset:")
display(oldest_10)

VALIDATION ANALYSIS

Birth year statistics:
count    659687.000000
mean       1954.320144
std          35.575371
min           4.000000
25%        1933.000000
50%        1961.000000
75%        1980.000000
max        2025.000000
Name: birthYear, dtype: float64

People born in 0s: 10

Oldest 10 people in dataset:


,primaryName,birthYear,deathYear,primaryProfession
737971,Lucio Anneo Seneca,4.0,65.0,writer
4295756,Philo Judaeus,20.0,50.0,miscellaneous
406802,Flavius Josephus,37.0,95.0,"writer,miscellaneous"
616071,Ovid,43.0,17.0,"writer,music_department"
3800194,Plutarch,46.0,122.0,writer
486269,Titus Livius,59.0,17.0,writer
14124112,Pliny the Younger,61.0,113.0,archive_footage
3805036,Suetonius,69.0,140.0,writer
518247,Publius Vergilius Maro,70.0,19.0,writer
3779892,Appian,95.0,165.0,writer


### Questions 5 & 6 - Answer

**Is the birth date (year 4) correct?**

**YES, it's correct!**

**Reasoning:**


1. BirthYear fits the dataset's distribution
    Year 4 is not an isolated outlier; other very early birth years also appear in the dataset.

2. Credits are logically consistent
    The person is credited as writer for modern films. IMDb commonly lists historical authors as writers when their works are adapted—this is normal behavior in the dataset.

2. No contradiction inside IMDb data
    The titles linked to this person exist and their release years do not conflict with the ancient birth date.

**Conclusion**
Based solely on IMDb data, the birth year 4 is valid and internally consistent.




## Question 7: What is the most recent date of birth?

In [14]:
# Find the most recent (latest) birth year
latest_birth_year = df_names['birthYear'].max()

# Get the person(s) with this birth year
latest_persons = df_names[df_names['birthYear'] == latest_birth_year]

print(f"Most recent year of birth: {int(latest_birth_year)}")
print(f"\nNumber of people born in {int(latest_birth_year)}: {len(latest_persons):,}")
print(f"\nSample of people with this birth year (showing up to 10):")
display(latest_persons.head(10)[['nconst', 'primaryName', 'birthYear', 'primaryProfession', 'knownForTitles']])

# Calculate current age
current_year = datetime.now().year
current_age = current_year - int(latest_birth_year)
print(f"\nCurrent age of people born in {int(latest_birth_year)}: {current_age} years old")

Most recent year of birth: 2025

Number of people born in 2025: 2

Sample of people with this birth year (showing up to 10):


,nconst,primaryName,birthYear,primaryProfession,knownForTitles
7063699,nm16784939,Kyrah Ivy Jackson,2025.0,actress,NaN
11502016,nm5642311,Chase Ramsey,2025.0,"actor,director,writer","tt17505010,tt14715170,tt4236770,tt17062324"



Current age of people born in 2025: 0 years old


### Question 7 - Answer

**Most recent date of birth:** 2025

This represents the youngest people in the IMDB database. Note that very recent birth years likely represent child actors or other young performers who have already appeared in credited works.

## Question 8: What percentage of people do not have a listed date of birth?

In [15]:
# Calculate the percentage of people without a birth year
total_people = len(df_names)
missing_birth_year = df_names['birthYear'].isna().sum()
percentage_missing = (missing_birth_year / total_people) * 100

print(f"Total people in dataset: {total_people:,}")
print(f"People with birth year: {(total_people - missing_birth_year):,}")
print(f"People WITHOUT birth year: {missing_birth_year:,}")
print(f"\nPercentage without birth year: {percentage_missing:.2f}%")
print(f"Percentage with birth year: {(100 - percentage_missing):.2f}%")

# Additional analysis
print("\n" + "="*50)
print("Additional Birth Year Statistics")
print("="*50)

# Check death year availability
missing_death_year = df_names['deathYear'].isna().sum()
percentage_missing_death = (missing_death_year / total_people) * 100
print(f"\nPeople without death year: {missing_death_year:,} ({percentage_missing_death:.2f}%)")

# People with birth but no death (likely still alive or death not recorded)
birth_no_death = df_names[(df_names['birthYear'].notna()) & (df_names['deathYear'].isna())]
print(f"People with birth year but no death year: {len(birth_no_death):,}")

Total people in dataset: 14,901,541
People with birth year: 659,687
People WITHOUT birth year: 14,241,854

Percentage without birth year: 95.57%
Percentage with birth year: 4.43%

Additional Birth Year Statistics

People without death year: 14,649,924 (98.31%)
People with birth year but no death year: 423,808


### Question 8 - Answer

**Percentage without date of birth:** 95.57%

## Question 9: What is the length of the longest "short" after 1900?

In [16]:
# Filter for "short" titles after 1900
# Convert startYear and runtimeMinutes to numeric
df_titles['startYear'] = pd.to_numeric(df_titles['startYear'], errors='coerce')
df_titles['runtimeMinutes'] = pd.to_numeric(df_titles['runtimeMinutes'], errors='coerce')

# Filter: titleType = 'short' AND startYear > 1900 AND runtimeMinutes is not null
shorts_after_1900 = df_titles[
    (df_titles['titleType'] == 'short') & 
    (df_titles['startYear'] > 1900) & 
    (df_titles['runtimeMinutes'].notna())
]

print(f"Total 'short' titles after 1900: {len(shorts_after_1900):,}")
print(f"Shorts with runtime data: {shorts_after_1900['runtimeMinutes'].notna().sum():,}")

# Find the longest short
if len(shorts_after_1900) > 0:
    longest_runtime = shorts_after_1900['runtimeMinutes'].max()
    longest_shorts = shorts_after_1900[shorts_after_1900['runtimeMinutes'] == longest_runtime]
    
    print(f"\nLength of longest 'short' after 1900: {int(longest_runtime)} minutes")
    print(f"\nTitle(s) with this runtime:")
    display(longest_shorts[['tconst', 'primaryTitle', 'startYear', 'runtimeMinutes', 'genres']])
    
    # Show statistics
    print("\nRuntime statistics for shorts after 1900:")
    print(shorts_after_1900['runtimeMinutes'].describe())
else:
    print("No shorts found after 1900 with runtime data.")

Total 'short' titles after 1900: 692,548
Shorts with runtime data: 692,548

Length of longest 'short' after 1900: 1311 minutes

Title(s) with this runtime:


,tconst,primaryTitle,startYear,runtimeMinutes,genres
8520598,tt35509411,Our First Day,2025.0,1311.0,"Drama,Short"



Runtime statistics for shorts after 1900:
count    692548.000000
mean         13.091026
std           9.152509
min           0.000000
25%           6.000000
50%          11.000000
75%          18.000000
max        1311.000000
Name: runtimeMinutes, dtype: float64


### Question 9 - Answer

**Length of longest 'short' after 1900:** 1311 minutes

## Question 10: What is the length of the shortest "movie" after 1900?

In [17]:
# Filter for "movie" titles after 1900
# Filter: titleType = 'movie' AND startYear > 1900 AND runtimeMinutes is not null
movies_after_1900 = df_titles[
    (df_titles['titleType'] == 'movie') & 
    (df_titles['startYear'] > 1900) & 
    (df_titles['runtimeMinutes'].notna())
]

print(f"Total 'movie' titles after 1900: {len(movies_after_1900):,}")
print(f"Movies with runtime data: {movies_after_1900['runtimeMinutes'].notna().sum():,}")

# Find the shortest movie
if len(movies_after_1900) > 0:
    shortest_runtime = movies_after_1900['runtimeMinutes'].min()
    shortest_movies = movies_after_1900[movies_after_1900['runtimeMinutes'] == shortest_runtime]
    
    print(f"\nLength of shortest 'movie' after 1900: {int(shortest_runtime)} minutes")
    print(f"\nTitle(s) with this runtime:")
    display(shortest_movies[['tconst', 'primaryTitle', 'startYear', 'runtimeMinutes', 'genres']].head(10))
    
    # Show statistics
    print("\nRuntime statistics for movies after 1900:")
    print(movies_after_1900['runtimeMinutes'].describe())
    
    # Show distribution of very short movies
    very_short = movies_after_1900[movies_after_1900['runtimeMinutes'] < 30]
    print(f"\nNumber of movies under 30 minutes: {len(very_short):,}")
else:
    print("No movies found after 1900 with runtime data.")

Total 'movie' titles after 1900: 454,789
Movies with runtime data: 454,789

Length of shortest 'movie' after 1900: 1 minutes

Title(s) with this runtime:


,tconst,primaryTitle,startYear,runtimeMinutes,genres
24728,tt0025166,George White's Scandals,1934.0,1.0,"Comedy,Musical,Romance"
450322,tt0469119,Love Trap,2005.0,1.0,Drama
784826,tt0810779,Bound by Blood,2007.0,1.0,Action
821099,tt0848384,Nikkatsu on Parade,1930.0,1.0,Documentary
2548543,tt12893768,If I Die Tomorrow,2020.0,1.0,Documentary
6249626,tt26348770,Dancing Boy,2023.0,1.0,Documentary
7722633,tt32276067,Honest Vikky (Life Coach),2024.0,1.0,Adult



Runtime statistics for movies after 1900:
count    454789.000000
mean         89.710002
std         148.080350
min           1.000000
25%          73.000000
50%          89.000000
75%         100.000000
max       59460.000000
Name: runtimeMinutes, dtype: float64

Number of movies under 30 minutes: 1,386


### Question 10 - Answer

**Length of shortest 'movie' after 1900:** 1 minutes

## Question 11: List of all genres represented

In [18]:
print("Extracting all genres from dataset...\n")

# Filter out null values and split by comma
all_genres = df_titles['genres'].dropna()
genre_list = []

for genres in all_genres:
    # Split by comma and strip whitespace
    individual_genres = [g.strip() for g in str(genres).split(',')]
    genre_list.extend(individual_genres)

# Get unique genres and sort them
unique_genres = sorted(set(genre_list))

print(f"Total number of unique genres: {len(unique_genres)}")
print("\nComplete list of all genres:")
print("=" * 50)

for i, genre in enumerate(unique_genres, 1):
    print(f"{i:2d}. {genre}")

# Count frequency of each genre
print("\n" + "=" * 50)
print("Genre Frequency Analysis")
print("=" * 50)

from collections import Counter
genre_counts = Counter(genre_list)
top_10_genres = genre_counts.most_common(10)

print("\nTop 10 most common genres:")
for genre, count in top_10_genres:
    print(f"{genre:20s}: {count:,} titles")

Extracting all genres from dataset...

Total number of unique genres: 28

Complete list of all genres:
 1. Action
 2. Adult
 3. Adventure
 4. Animation
 5. Biography
 6. Comedy
 7. Crime
 8. Documentary
 9. Drama
10. Family
11. Fantasy
12. Film-Noir
13. Game-Show
14. History
15. Horror
16. Music
17. Musical
18. Mystery
19. News
20. Reality-TV
21. Romance
22. Sci-Fi
23. Short
24. Sport
25. Talk-Show
26. Thriller
27. War
28. Western

Genre Frequency Analysis

Top 10 most common genres:
Drama               : 3,397,777 titles
Comedy              : 2,338,642 titles
Talk-Show           : 1,500,346 titles
Short               : 1,283,189 titles
News                : 1,177,099 titles
Documentary         : 1,151,440 titles
Romance             : 1,124,984 titles
Family              : 878,811 titles
Reality-TV          : 676,913 titles
Animation           : 603,839 titles


### Question 11 - Answer

**list of all genres:** 28 genres

## Question 12: Highest rated comedy "movie" in the dataset

In [19]:
# Step 1: Filter for comedy movies
print("Finding highest rated comedy movie...\n")

# Filter titles: must be 'movie' and have 'Comedy' in genres
comedy_movies = df_titles[
    (df_titles['titleType'] == 'movie') & 
    (df_titles['genres'].notna()) &
    (df_titles['genres'].str.contains('Comedy', case=False, na=False))
]

print(f"Total comedy movies: {len(comedy_movies):,}")

# Step 2: Merge with ratings data
comedy_with_ratings = comedy_movies.merge(df_ratings, on='tconst', how='inner')

print(f"Comedy movies with ratings: {len(comedy_with_ratings):,}")

# Step 3: Sort by rating (desc) and then by votes (desc) for tiebreaking
comedy_with_ratings_sorted = comedy_with_ratings.sort_values(
    by=['averageRating', 'numVotes'], 
    ascending=[False, False]
)

# Get the highest rated comedy movie
highest_rated_comedy = comedy_with_ratings_sorted.iloc[0]

print("\n" + "=" * 70)
print("HIGHEST RATED COMEDY MOVIE")
print("=" * 70)
print(f"\nTitle: {highest_rated_comedy['primaryTitle']}")
print(f"Year: {int(highest_rated_comedy['startYear']) if pd.notna(highest_rated_comedy['startYear']) else 'N/A'}")
print(f"Rating: {highest_rated_comedy['averageRating']:.1f}/10")
print(f"Number of Votes: {int(highest_rated_comedy['numVotes']):,}")
print(f"Runtime: {int(highest_rated_comedy['runtimeMinutes']) if pd.notna(highest_rated_comedy['runtimeMinutes']) else 'N/A'} minutes")
print(f"Genres: {highest_rated_comedy['genres']}")
print(f"IMDB ID: {highest_rated_comedy['tconst']}")

# Store for later use
best_comedy_id = highest_rated_comedy['tconst']
best_comedy_title = highest_rated_comedy['primaryTitle']

# Show top 10 for context
print("\n" + "=" * 70)
print("Top 10 Highest Rated Comedy Movies")
print("=" * 70)
display(comedy_with_ratings_sorted.head(10)[[
    'primaryTitle', 'startYear', 'averageRating', 'numVotes', 'genres'
]])

Finding highest rated comedy movie...

Total comedy movies: 122,129
Comedy movies with ratings: 80,769

HIGHEST RATED COMEDY MOVIE

Title: Bob vs. Society
Year: 2009
Rating: 10.0/10
Number of Votes: 12
Runtime: 90 minutes
Genres: Comedy
IMDB ID: tt1423343

Top 10 Highest Rated Comedy Movies


,primaryTitle,startYear,averageRating,numVotes,genres
49238,Bob vs. Society,2009.0,10.0,12,Comedy
73413,Happy Lucky Golden Tofu Panda Dragon Good Time...,2017.0,10.0,11,"Comedy,Documentary,Music"
51339,Reck and Ima,2013.0,10.0,5,Comedy
69482,Kamen batyrov,2025.0,10.0,5,"Comedy,Fantasy,Mystery"
61939,Rabb Di Awaaz,2023.0,9.9,22,"Comedy,Drama"
64574,The ReWrite.,2025.0,9.9,22,"Comedy,Drama"
62604,Shubh Lagan Muhurt,2023.0,9.9,14,Comedy
61152,Zucchini,2025.0,9.9,8,"Comedy,Romance"
67475,Borderline Forever,2021.0,9.8,93,"Adventure,Comedy,Music"
68949,Em Queda-Livre,2024.0,9.8,41,Comedy


### Question 12 - Answer

**Highest rated comedy movie:** Bob vs. Society

## Question 13: Who was the director of the movie?

In [20]:
# Find the director(s) of the highest rated comedy movie
print(f"Finding director(s) of '{best_comedy_title}'...\n")

# Look up in crew data
movie_crew = df_crew[df_crew['tconst'] == best_comedy_id]

if len(movie_crew) > 0:
    directors_ids = movie_crew.iloc[0]['directors']
    
    if pd.notna(directors_ids) and directors_ids != '':
        # Split director IDs (may be multiple)
        director_id_list = directors_ids.split(',')
        
        print(f"Director ID(s): {directors_ids}")
        print(f"Number of directors: {len(director_id_list)}\n")
        
        # Look up director names
        directors_info = df_names[df_names['nconst'].isin(director_id_list)]
        
        print("=" * 70)
        print(f"DIRECTOR(S) OF '{best_comedy_title}'")
        print("=" * 70)
        
        if len(directors_info) > 0:
            for idx, director in directors_info.iterrows():
                print(f"\nName: {director['primaryName']}")
                print(f"ID: {director['nconst']}")
                print(f"Birth Year: {int(director['birthYear']) if pd.notna(director['birthYear']) else 'N/A'}")
                print(f"Professions: {director['primaryProfession']}")
                print(f"Known for: {director['knownForTitles']}")
            
            # Display in table format
            print("\nDirector Information (Table):")
            display(directors_info[['nconst', 'primaryName', 'birthYear', 'deathYear', 'primaryProfession']])
        else:
            print("Director information not found in name.basics dataset.")
            print(f"Director ID(s): {directors_ids}")
    else:
        print("No director information available for this movie.")
else:
    print(f"Crew information not found for {best_comedy_id}")

Finding director(s) of 'Bob vs. Society'...

Director ID(s): nm2301250
Number of directors: 1

DIRECTOR(S) OF 'Bob vs. Society'

Name: David Pring-Mill
ID: nm2301250
Birth Year: N/A
Professions: production_department,writer,director
Known for: tt1423343,tt5948938,tt1545007,tt3271334

Director Information (Table):


,nconst,primaryName,birthYear,deathYear,primaryProfession
8568010,nm2301250,David Pring-Mill,NaN,NaN,"production_department,writer,director"


### Question 13 - Answer

**Director(s) of the highest rated comedy movie:** David Pring-Mill

## Question 14: List alternate titles for the movie

In [21]:
# Find alternate titles (akas) for the highest rated comedy movie
print(f"Finding alternate titles for '{best_comedy_title}'...\n")

# Look up in akas data
movie_akas = df_akas[df_akas['titleId'] == best_comedy_id]

print("=" * 70)
print(f"ALTERNATE TITLES FOR '{best_comedy_title}'")
print("=" * 70)

if len(movie_akas) > 0:
    print(f"\nTotal alternate titles found: {len(movie_akas)}\n")
    
    # Display all alternate titles
    display(movie_akas[[
        'title', 'region', 'language', 'types', 'attributes', 'isOriginalTitle'
    ]])
    
    # Summary by region
    print("\n" + "=" * 70)
    print("Summary by Region")
    print("=" * 70)
    region_counts = movie_akas['region'].value_counts()
    print(f"\nNumber of different regions: {len(region_counts)}")
    print("\nTop 10 regions with most alternate titles:")
    print(region_counts.head(10))
    
    # Show original title if available
    original = movie_akas[movie_akas['isOriginalTitle'] == 1]
    if len(original) > 0:
        print("\n" + "=" * 70)
        print("Original Title(s)")
        print("=" * 70)
        display(original[['title', 'region', 'language']])
    
else:
    print("\nNo alternate titles found for this movie in the database.")
    print("This could mean:")
    print("  - The movie was only released with its primary title")
    print("  - Alternate title data is not available")
    print("  - The movie has limited international distribution")

Finding alternate titles for 'Bob vs. Society'...

ALTERNATE TITLES FOR 'Bob vs. Society'

Total alternate titles found: 3



,title,region,language,types,attributes,isOriginalTitle
15197893,Bob vs. Society,NaN,NaN,original,NaN,1
15197894,Bob vs. Society,GB,NaN,imdbDisplay,NaN,0
15197895,Bob vs. Society,US,NaN,imdbDisplay,NaN,0



Summary by Region

Number of different regions: 2

Top 10 regions with most alternate titles:
region
GB    1
US    1
Name: count, dtype: int64

Original Title(s)


,title,region,language
15197893,Bob vs. Society,NaN,NaN


### Question 14 - Answer

**Alternate titles:** 3 titles